### ISSUE TITLE
Promote scratch images as official release images

### ISSUE THREAD

**Issue by rturner3** [spiffe/spire#3545]
#1972 introduced build targets for scratch images that produce images with only the artifacts necessary to run SPIRE components.  The stripped-down images provide a safer security posture for users who depend on SPIRE release images, such that only the required software and configuration is present in the container.

We have been publishing these scratch release images to GitHub Packages as of v1.1.0, however we have not advertised them in any project documentation or release notes. We should promote these scratch images as the official release images and eventually retire the bulkier images that we still publish to Google Container Registry.

Sub-tasks:

- [x] Replace all readiness/liveness probes in examples/tests that rely on non-SPIRE binaries being present in the container
    - https://github.com/spiffe/spire-tutorials/pull/97
- [x] Replace all references in tests to GCR images with GHCR images in SPIRE repository
    - https://github.com/spiffe/spire/pull/3636
- [x] Replace all references to GCR images with GHCR images in [spiffe/spire-tutorials](https://www.github.com/spiffe/spire-tutorials), fixing examples as required
    - https://github.com/spiffe/spire-tutorials/pull/98
- [x] Replace all references to GCR images with GHCR images in [spiffe/spire-examples](https://www.github.com/spiffe/spire-examples), fixing examples as required
    - https://github.com/spiffe/spire-examples/pull/53
- [x] Replace all references to GCR images with GHCR images in [spiffe/tornjak](https://www.github.com/spiffe/tornjak)
    - https://github.com/spiffe/tornjak/pull/97
- [x] Document existence of GHCR release images in SPIRE README.md
    - https://github.com/spiffe/spire/pull/3641
- [x] Evaluate issue with running scratch images from GHCR on Apple Silicon Macs #3749
- [x] Add default non-root user to scratch images
    - https://github.com/spiffe/spire/pull/3811
- [x] Add guide for moving to scratch images with known implications for containers running as non-root user
    - https://github.com/spiffe/spire/pull/3811

**Comment by marcofranssen** [comment]
There is one issue with that. Readiness probes will fail because the required binaries are not there. Is there another way to do readiness probes on these containers? If yes all the examples out there have to be updated.

```yaml
          readinessProbe:
            exec:
              command: ["/bin/ps", "aux", "|", "grep", "oidc-discovery-provider -config /run/spire/oidc/config/oidc-discovery-provider.conf"]
            initialDelaySeconds: 5
            periodSeconds: 5

```

**Comment by azdagron** [comment]
Yikes. That ps based readiness probe doesn't seem like it should have been used in the first place, as K8s will already be aware if that process goes down. We should definitely drop those (or convert to the http probes).

**Comment by rturner3** [comment]
I tried to use the OIDC Discovery Provider HTTP health check endpoints in Kubernetes examples, but I found that it doesn't work when not using the host network. See #3629.

I am inclined to just remove the OIDC Discovery Provider readiness probes for now since the fix requires code changes in OIDC Discovery Provider.

**Comment by rturner3** [comment]
Removing `ps`-based readiness probes in https://github.com/spiffe/spire-tutorials/pull/97

**Comment by rturner3** [comment]
Replacing references in spiffe/spire-examples: https://github.com/spiffe/spire-examples/pull/53
Replacing references in spiffe/spire-tutorials: https://github.com/spiffe/spire-tutorials/pull/98

**Comment by rturner3** [comment]
Replacing references in spiffe/tornjak: https://github.com/spiffe/tornjak/pull/97

**Comment by marcofranssen** [comment]
> I tried to use the OIDC Discovery Provider HTTP health check endpoints in Kubernetes examples, but I found that it doesn't work when not using the host network. See #3629.
> 
> I am inclined to just remove the OIDC Discovery Provider readiness probes for now since the fix requires code changes in OIDC Discovery Provider.

The fix for this is already on main #3580

**Comment by rturner3** [comment]
> > I tried to use the OIDC Discovery Provider HTTP health check endpoints in Kubernetes examples, but I found that it doesn't work when not using the host network. See #3629.
> > I am inclined to just remove the OIDC Discovery Provider readiness probes for now since the fix requires code changes in OIDC Discovery Provider.
> 
> The fix for this is already on main #3580

Oh excellent, I missed that. Makes a lot of sense now looking at the PR why it is broken with the latest released version.

**Comment by rturner3** [comment]
There is one noteworthy caveat I've discovered so far with the scratch images - it is not straightforward to run the container as a non-root user. By default in a scratch image the root directory of the filesystem `/` is owned by root and is not writeable by any other user. This presents problems with SPIRE because there are several directories on disk that spire-server and spire-agent will attempt to create at runtime if they are missing from the filesystem (e.g. Unix socket directory, data directory). These operations will fail by default if the container is not running as root.

This is not ideal from a security perspective since running as a root user can grant SPIRE unnecessary privileges on the host. https://github.com/spiffe/spire/issues/1862 shows that the community has a desire to run SPIRE images as a non-root user in some environments.

There is a workaround for this filesystem permission issue at build time of the image by doing the following:

- Creating a dedicated user and group  (i.e. `/etc/passwd` and `/etc/group` files) in the container using tools available in the builder image and copying these files to the scratch image
- Creating the necessary directories in the container with the correct ownership and permissions using `ADD` Dockerfile directives in the builder image and `COPY` Dockerfile directives to add the directories to the scratch image.

These image configurations can be set up in the scratch images with some default non-root user + group for all the configuration fields that have default path values. However this now means there is a tight coupling between the image building process and the configuration of SPIRE components. The impact for users that wish to run these images as a non-root user is that they will need to either:
- Accept all the default paths and add volume mounts with correct permissions/ownership for any other paths that SPIRE needs to read from/write to
- Create their own custom image build that sources the release scratch images and set up this configuration themselves

This should be documented in a new guide that users can refer to when considering whether to move to the new release images based on scratch. I've added this as a sub-task in the original issue text.
